### ISSUE TITLE
Crash when calling collect on Worker running closures.

### ISSUE THREAD

**Issue by wingzx3** [krakjoe/pthreads#812]
Environment
==========

PHP 7.2.1 (cli) (built: Jan  4 2018 04:29:12) ( ZTS MSVC15 (Visual C++ 2017) x86 )
pthreads = 3.1.7dev
Windows 10 Version 10.0.16299 Build 16299

Summary
======

Run this code for a while.  For me it crashes in about 7 seconds.  If you comment out the ``($this->closure)();`` it'll run forever.

Reproducing Code
==============

```
class TestClosure extends Threaded implements Collectable {
  protected $closure;

  function __construct( $closure ) {
     $this->closure = $closure;
  }

  public function run() {
    ($this->closure)();
  }

};

$worker = new Worker();
$worker->start();
while (true) {
  $function = new TestClosure(function() {
    return 1;
  });
  $worker->stack($function);
  while($worker->collect());
  sleep(1);
}


```

Expected Output
=============

This code should not crash.  It should run forever.

Actual Output
============

Exited with code=3221226505 in 7.084 seconds

**Comment by dktapps** [comment]
I did some dumb debugging and it appears that the crash is occurring because the function is found in the `resolve` hashtable [here](https://github.com/krakjoe/pthreads/blob/master/src/copy.h#L382), but the value retrieved from the ht is garbage.

```
stacking task
function name: {closure} (refcount 2) ... created copy: name {closure} (refcount 1)
stacking task
function name: {closure} (refcount 2) ... created copy: name {closure} (refcount 1)
stacking task
function name: {closure} (refcount 2) ... created copy: name {closure} (refcount 1)
stacking task
function name: {closure} (refcount 2) ... created copy: name {closure} (refcount 1)
stacking task
function name: {closure} (refcount 2) ... created copy: name {closure} (refcount 1)
stacking task
function name: {closure} (refcount 2) ... created copy: name {closure} (refcount 1)
stacking task
function name: {closure} (refcount 2) ... got existing copy name: testclosure (refcount -335536079)
```
followed by segfault.

It looks like the memory allocated for the function stored in the resolve HT is later freed internally and allocated for something else, resulting in the memory being reused for something else.

Initializing [this refcount](https://github.com/krakjoe/pthreads/blob/master/src/copy.h#L344) to 2 instead of 1 stops the segfault when testing locally, but I do not know if this is a correct fix.

**Comment by SirSnyder** [comment]
I guess we have to add a function ref at https://github.com/krakjoe/pthreads/blob/master/src/store.c#L730, if the function was added to `resolve`. Tomorrow, not yet.

**Comment by dktapps** [comment]
A simple (stupid) solution would be to just get rid of function storing in the `resolve` hashtable completely. Apart from closures, functions will only ever get copied once per thread (in theory) - the only time that is useful is for closures, and yet here is this bug.

Additionally, the `resolve` hashtable is _only_ used for functions now as far as I can see, so it could be removed entirely.

**Comment by dktapps** [comment]
I tried to fix this by adding an extra ref to everything added to the resolve HT and then clean up those refs before HT destroy, but I ran into all kinds of other strange issues (inconsistent HT asserts and such) so I let go of trying to fix this myself. Learning curve, I'm (relatively) a noob in this field ¯\\\_(ツ)_/¯

Also I thought this bug would affect all the things, but it appears the problem only appears with closures (since normal functions don't get freed until script end).

**Comment by SirSnyder** [comment]
I thought about this topic the last few days with no satisfying result. I would like to have a solution that detects changes of the stored closures. But even if that got solved, the not insignifcant problem for long running applications of remaining/leaking entries in `HT(resolve)` still exists. In the end a halfway working solution was complicated and error prone. For now I think it's probably better to bypass the `HT(resolve)` in case of closures.

**Comment by SirSnyder** [comment]
Fixed with e25675de
